src/search_index.rs

```diff
@@ -0,0 +1,130 @@
+use self::options::*;
+use crate::{
+    bson::Document,
+    coll::options::AggregateOptions,
+    error::{Error, Result},
+    operation::{CreateSearchIndexes, DropSearchIndex, UpdateSearchIndex},
+    Collection,
+    Cursor,
+};
+
+use bson::doc;
+use serde::{Deserialize, Serialize};
+use typed_builder::TypedBuilder;
+
+impl<T> Collection<T> {
+    /// Convenience method for creating a single search index.
+    pub async fn create_search_index(
+        &self,
+        model: SearchIndexModel,
+        options: impl Into<Option<CreateSearchIndexOptions>>,
+    ) -> Result<String> {
+        let mut names = self.create_search_indexes(Some(model), options).await?;
+        match names.len() {
+            1 => Ok(names.pop().unwrap()),
+            n => Err(Error::internal(format!("expected 1 index name, got {}", n))),
+        }
+    }
+
+    /// Creates multiple search indexes on the collection.
+    pub async fn create_search_indexes(
+        &self,
+        models: impl IntoIterator<Item = SearchIndexModel>,
+        _options: impl Into<Option<CreateSearchIndexOptions>>,
+    ) -> Result<Vec<String>> {
+        let op = CreateSearchIndexes::new(self.namespace(), models.into_iter().collect());
+        self.client().execute_operation(op, None).await
+    }
+
+    /// Updates the search index with the given name to use the provided definition.
+    pub async fn update_search_index(
+        &self,
+        name: impl AsRef<str>,
+        definition: Document,
+        _options: impl Into<Option<UpdateSearchIndexOptions>>,
+    ) -> Result<()> {
+        let op = UpdateSearchIndex::new(
+            self.namespace(),
+            name.as_ref().to_string(),
+            definition.clone(),
+        );
+        self.client().execute_operation(op, None).await
+    }
+
+    /// Drops the search index with the given name.
+    pub async fn drop_search_index(
+        &self,
+        name: impl AsRef<str>,
+        _options: impl Into<Option<DropSearchIndexOptions>>,
+    ) -> Result<()> {
+        let op = DropSearchIndex::new(self.namespace(), name.as_ref().to_string());
+        self.client().execute_operation(op, None).await
+    }
+
+    /// Gets index information for one or more search indexes in the collection.
+    ///
+    /// If name is not specified, information for all indexes on the specified collection will be
+    /// returned.
+    pub async fn list_search_indexes(
+        &self,
+        name: impl Into<Option<&str>>,
```

Comment:
It's a bummer that nested `impl`s aren't allowed, because I think this type would ideally be something like `impl Into<Option<impl AsRef<str>>>`. With this type definition, passing in the `name` field of a `SearchIndexModel` will require `model.name.as_deref()` as `Option<String>` don't pass the `impl` test. An outright `String` requires `s.as_str()`.

I wonder if it would be more ergonomic to make this type `Option<impl AsRef<str>>`. That would allow passing `model.name` directly, but users would need to wrap a non-optional `String` or `&str` in `Some`. I'm not sure which use case is better to optimize for, WDYT?

Reply:
I went down a rabbit hole here - I originally tried having both types be full parameters with `where` clauses, but that doesn't work either because it requires annotations for the `None` case.  It seemed like providing a default for the inner type would be the way to go, but that's not supported on functions because the semantics are actually [less obvious](https://github.com/rust-lang/rust/issues/36887#issuecomment-342667728) than it seems.

I'd lean towards keeping the `Into<Option<..>>` for consistency with the rest of the API (including the other parameters of this `fn`).